### PR TITLE
[Snackbar] Allow the delegate to override accessibilityViewIsModal

### LINKED ISSUE #5879: [Snackbar] MDCSnackbar text in VoiceOver mode is not fully read
This was filed as an internal issue. If you are a Googler, please visit [b/120225668](http://b/120225668) for more details.

---
#### Internal data
- Associated internal bug: [b/120225668](http://b/120225668)

### PR DESCRIPTION
**Pull request by fluxminds**
Enabling client's ability to overwrite the accessibilityViewIsModal in snackbar view by the delegate method.

closes #5879

### CODE REVIEW

**Comment by fluxminds** [comment]
Discussed offline with Rob and Randall, we agreed with moving to another approach, which allows clients to overwrite the accessibilityViewIsModal in delegate to affect the manager's behavior. Will push some changes according to the conversation.

**Comment by fluxminds** [comment]
@romoore Changes have been made according to our conversation offline.

File: components/Snackbar/src/MDCSnackbarManager.m [hunk 1]
```diff
@@ -230,28 +230,45 @@ - (void)displaySnackbarViewForMessage:(MDCSnackbarMessage *)message {
...
   // only if the user isn't running VoiceOver.
+  __weak MDCSnackbarManagerInternal *weakSelf = self;
   [self.overlayView
       showSnackbarView:snackbarView
               animated:YES
             completion:^{
-              if ([self isSnackbarTransient:snackbarView]) {
+              MDCSnackbarManagerInternal *strongSelf = weakSelf;
+              if (!accessibilityViewIsModal &&
+                  [strongSelf isSnackbarTransient:snackbarView]) {
```
[the remark this answers is not in the document]
Comment (romoore) [reply components/Snackbar/src/MDCSnackbarManager.m]: Looks like this is still outstanding?

File: components/Snackbar/src/MDCSnackbarManager.m [hunk 2]
```diff
@@ -286,26 +289,29 @@ - (void)hideSnackbarViewReally:(MDCSnackbarMessageView *)snackbarView
...
+                                       // announced (layout was not reported as changed) so there is
+                                       // no need to post a layout change here.
+                                       if (strongSelf.overlayView.accessibilityViewIsModal ||
+                                           ![strongSelf isSnackbarTransient:snackbarView]) {
                                          UIAccessibilityPostNotification(
                                              UIAccessibilityLayoutChangedNotification, nil);
                                        }
 
-                                       self.currentSnackbar = nil;
+                                       strongSelf.currentSnackbar = nil;
```
Comment (romoore) [inline components/Snackbar/src/MDCSnackbarManager.m]: Are these strong/weak changes required for this PR? Was there previously a leak in the API that we missed?   If not, could they please be moved to a separate PR?  If they are required, please add to the PR summary.

**Comment by romoore** [comment]
Looks like the summary is outdated.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m [hunk 3]
```diff
@@ -123,7 +123,42 @@ - (void)testAccessibilityHintSetFromSnackbarMessageProperty {
...
+  // When
+  [self.manager showMessage:self.message];
+  XCTestExpectation *expectation = [self expectationWithDescription:@"completed"];
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [expectation fulfill];
+  });
+  [self waitForExpectationsWithTimeout:3 handler:nil];
+
+  // Then
+  XCTAssertFalse(self.manager.internalManager.overlayView.accessibilityViewIsModal);
```
Comment (romoore) [inline components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:142]: Shouldn't this be `XCTAssertTrue` given the test name? "ViewIsModalByDefaultWithActions"?

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m [hunk 4]
```diff
@@ -123,7 +123,42 @@ - (void)testAccessibilityHintSetFromSnackbarMessageProperty {
...
```
Comment (romoore) [inline components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:158]: This should also be `XCTAssertTrue` given the test name.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m [hunk 5]
```diff
@@ -143,8 +178,9 @@ - (void)testSnackbarSetAccessibiltyViewIsModalForActionSnacbars {
   XCTAssertTrue(self.manager.internalManager.overlayView.accessibilityViewIsModal);
 }
 
-- (void)testSnackbarAccessibiltyViewIsModalShouldBeNoWithNoActions {
+- (void)testSnackbarAccessibiltyViewIsModalYesWithNoActions {
   // Given
+  self.manager.internalManager.isVoiceOverRunningOverride = YES;
```
Comment (romoore) [inline components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:202]: Why is the override required now for the test? This looks like a regression/change in expected behavior.

**Review by romoore — COMMENTED** [review]
I have some concerns about the unexpected functionality changes that the tests are exposing. Let's go through this in person.

File: components/Snackbar/src/MDCSnackbarManager.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (fluxminds) [reply components/Snackbar/src/MDCSnackbarManager.m]: Seems a clang format update issue. Will see if the latest commit still complain it or not and fix it accordingly.

File: components/Snackbar/src/MDCSnackbarManager.m (hunk 2 above)
Comment (fluxminds) [reply components/Snackbar/src/MDCSnackbarManager.m]: Not necessary for this PR. Reverted in 6d014757978339e18acd6c96bcfb21872eaf295b.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m (hunk 3 above)
Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:142]: Followed up offline. Renamed the method in 6d014757978339e18acd6c96bcfb21872eaf295b to make it clear.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m (hunk 4 above)
Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:158]: Same. Renamed the method in 6d014757978339e18acd6c96bcfb21872eaf295b to make it clear.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m (hunk 5 above)
Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:202]: Addressed in 6d014757978339e18acd6c96bcfb21872eaf295b. Added another test method to branch the case.

File: components/Snackbar/src/MDCSnackbarManager.m [hunk 6]
```diff
@@ -241,11 +242,11 @@ - (void)displaySnackbarViewForMessage:(MDCSnackbarMessage *)message {
       showSnackbarView:snackbarView
               animated:YES
             completion:^{
-              if ([self isSnackbarTransient:snackbarView]) {
+              if (!snackbarView.accessibilityViewIsModal &&
```
Comment (romoore) [inline components/Snackbar/src/MDCSnackbarManager.m]: Tip: It would be nice for this conditional to match the one in the dismiss block below.  I would consider swapping the if/else blocks here so you can use the same set of logical conditions.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m [hunk 7]
```diff
@@ -123,7 +123,42 @@ - (void)testAccessibilityHintSetFromSnackbarMessageProperty {
                         self.message.accessibilityHint);
 }
 
-- (void)testSnackbarSetAccessibiltyViewIsModalForActionSnacbars {
+- (void)testSnackbarAccessibiltyViewIsModalByDefaultToNoWithActions {
```
Comment (romoore) [inline components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m]: I think we spoke about the naming here. I believe we settled on something like `testSnackbarAccessibiltyViewIsModalDefaultsToNoWithActions`?

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m [hunk 8]
```diff
@@ -143,8 +178,28 @@ - (void)testSnackbarSetAccessibiltyViewIsModalForActionSnacbars {
...
+  // When
+  [self.manager showMessage:self.message];
+  XCTestExpectation *expectation = [self expectationWithDescription:@"completed"];
+  dispatch_async(dispatch_get_main_queue(), ^{
+    [expectation fulfill];
+  });
+  [self waitForExpectationsWithTimeout:3 handler:nil];
+
+  // Then
+  XCTAssertFalse(self.manager.internalManager.overlayView.accessibilityViewIsModal);
```
Comment (romoore) [inline components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:197]: If the name is "ViewIsModal**Yes**With..", shouldn't this be an XCTAssertTrue?

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m [hunk 9]
```diff
@@ -159,13 +214,14 @@ - (void)testSnackbarAccessibiltyViewIsModalShouldBeNoWithNoActions {
   XCTAssertFalse(self.manager.internalManager.overlayView.accessibilityViewIsModal);
```
Comment (romoore) [inline components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:214]: Same here, shouldn't this be an `XCTAssertTrue` given that it's supposed to be "YES"?

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m [hunk 10]
```diff
@@ -159,13 +214,14 @@ - (void)testSnackbarAccessibiltyViewIsModalShouldBeNoWithNoActions {
...
-- (void)testSnackbarSetAccessibiltyViewIsModalShouldBeNoForActionSnacbarsWhenManagerIsNo {
+- (void)testSnackbarAccessibiltyViewIsModalWithSnackbarViewOverwriteWithActions {
   // Given
+  self.delegate.shouldSetSnackbarViewAccessibilityViewIsModal = YES;
   self.manager.internalManager.isVoiceOverRunningOverride = YES;
   MDCSnackbarMessageAction *action = [[MDCSnackbarMessageAction alloc] init];
   action.title = @"Tap Me";
   self.message.action = action;
-  self.manager.shouldEnableAccessibilityViewIsModal = NO;
+  self.message.accessibilityViewIsModal = YES;
```
Comment (romoore) [inline components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m]: If this work, then shouldn't this be the preferred mechanism for clients, rather than relying on the delegate?

Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m]: This line needs to be removed.  And yes, `shouldEnableAccessibilityViewIsModal` should be the preferred one to use.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m (hunk 9 above)
Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:214]: Addressed in a5ddbbd01590490c2cebea9e302ad8fbdd774c28.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m (hunk 8 above)
Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m:197]: Followed up offline. Addressed in a5ddbbd01590490c2cebea9e302ad8fbdd774c28.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m (hunk 7 above)
Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m]: Addressed in a5ddbbd01590490c2cebea9e302ad8fbdd774c28

File: components/Snackbar/src/MDCSnackbarManager.m (hunk 6 above)
Comment (fluxminds) [reply components/Snackbar/src/MDCSnackbarManager.m]: Sure. Addressed in a5ddbbd01590490c2cebea9e302ad8fbdd774c28.

File: components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m (hunk 10 above)
Comment (fluxminds) [reply components/Snackbar/tests/unit/MDCSnackbarMessageViewTests.m]: Addressed in a5ddbbd01590490c2cebea9e302ad8fbdd774c28

**Review by romoore — APPROVED** [review]

**Comment by romoore** [comment]
I tweaked the subject a bit. I think this is a bit more clear.

**Comment by fluxminds** [comment]
@romoore Thanks for helping tweaking the PR subject.
